[ClangImporter] Anonymous enums used as types can't be imported as Int

jrose-apple opened the pull request:
Normally we treat anonymous enums as defining constants, and since we don't know how those constants are going to be used we import them as Int when we know we can do so consistently across all platforms. But this isn't the right thing to do if the anonymous enum is being used as an ad-hoc type for a variable or field.

(This may *still* not be the right thing to do—maybe we want to import it with a made-up name like we do anonymous fields instead. But at least this way any such variables or fields become usable, when they previously crashed IRGen.)

[SR-2511](https://bugs.swift.org/browse/SR-2511) / rdar://problem/35702809

jrose-apple commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -812,7 +812,7 @@ namespace {
         // they fit in an Int32. If not, this mapping isn't guaranteed to be
         // consistent for all platforms we care about.
-        if (!clangDecl->isFixed() &&
+        if (!clangDecl->isFixed() && clangDecl->isFreeStanding() &&
```
I asked on the Clang list whether `isFreeStanding` or `isEmbeddedInDeclarator` is the appropriate query here; my initial impression is that they do the same thing.

slavapestov replied:
If they're the same shouldn't one of them be deleted?

jrose-apple replied:
I'm assuming they're *not* the same and I'm just missing something, but if they *are* the same I bet someone will ask me to do that.

rjmccall replied:
Yes, I'm not sure why they wouldn't just be exact opposites.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

aschwaighofer reviewed (approved):
This looks good to me but I am not a clang expert so I don't know what the right query to use is here. ``isFreeStanding`` seems reasonable to me.

jrose-apple committed:
[ClangImporter] Anonymous enums used as types can't be imported as Int

Normally we treat anonymous enums as defining constants, and since we
don't know how those constants are going to be used we import them as
Int when we know we can do so consistently across all platforms. But
this isn't the right thing to do if the anonymous enum is being used
as an ad-hoc type for a variable or field.

(This may /still/ not be the right thing to do---maybe we want to
import it with a made-up name like we do anonymous fields instead.
But at least this way any such variables or fields become usable,
when they previously crashed IRGen.)

https://bugs.swift.org/browse/SR-2511

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
No one got back to me on the Clang side, so I'm going to merge this.
